crates/steel/src/event.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2025 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use alloy_rpc_types::{Topic, ValueOrArray};
+
+use crate::{state::WrapStateDb, EvmBlockHeader, EvmDatabase, GuestEvmEnv};
+use alloy_primitives::{Address, Log, Sealed};
+use alloy_rpc_types::Filter;
+use alloy_sol_types::SolEvent;
+use std::marker::PhantomData;
+
+/// Represents an Ethereum event query.
+///
+/// ### Usage
+/// - **Preflight calls on the Host:** To prepare calls on the host environment and build the
+///   necessary proof, use [Event::preflight].
+/// - **Calls in the Guest:** To initialize the contract in the guest environment, use [Event::new].
+///
+/// See [Contract] for more details.
+///
+/// ### Examples
+/// ```rust,no_run
+/// # use risc0_steel::{ethereum::EthEvmEnv, Event};
+/// # use alloy_primitives::address;
+/// # use alloy_sol_types::sol;
+///
+/// # #[tokio::main(flavor = "current_thread")]
+/// # async fn main() -> anyhow::Result<()> {
+/// let contract_address = address!("dAC17F958D2ee523a2206206994597C13D831ec7");
+/// sol! {
+/// #     #[derive(Debug)]
+///     interface IERC20 {
+///         event Transfer(address indexed from, address indexed to, uint256 value);
+///     }
+/// }
+///
+/// // Host:
+/// let url = "https://ethereum-rpc.publicnode.com".parse()?;
+/// let mut env = EthEvmEnv::builder().rpc(url).build().await?;
+/// let event = Event::preflight::<IERC20::Transfer>(&mut env).address(contract_address);
+/// event.query().await?;
+///
+/// let evm_input = env.into_input().await?;
+///
+/// // Guest:
+/// let env = evm_input.into_env();
+/// let event = Event::new::<IERC20::Transfer>(&env).address(contract_address);
+/// let logs = event.query();
+/// # dbg!(logs);
+///
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [Contract]: crate::Contract
+pub struct Event<S, E> {
+    filter: Filter,
+    env: E,
+    phantom: PhantomData<S>,
+}
+
+impl<H: EvmBlockHeader> Event<(), &GuestEvmEnv<H>> {
+    /// Constructor for executing an event query for a specific Solidity event.
+    pub fn new<S: SolEvent>(env: &GuestEvmEnv<H>) -> Event<S, &GuestEvmEnv<H>> {
+        Event {
+            filter: event_filter::<S, H>(env.header()),
+            env,
+            phantom: PhantomData,
+        }
+    }
+}
```

Comment:
Implementing this for `Event<(), E>` makes it possible to initialize with
```rust
Event::new::<IERC20::Transfer>(&env)
```

Instead of having to also specify `E`.

In general I decided to assign a fixed SolEvent to a `Steel`. This allows the corresponding Event::query() to return the actual events of the correct type.
An alternative would be to specify an arbitrary filter in the query step and leave all casting and conversion to the user.